core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -0,0 +1,291 @@
+
+import android.bluetooth.BluetoothDevice
+import android.bluetooth.BluetoothGattCharacteristic
+import android.bluetooth.BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
+import android.bluetooth.BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
+import android.bluetooth.BluetoothGattDescriptor
+import android.bluetooth.BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE
+import android.bluetooth.BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
+import android.content.Context
+import android.util.Log
+import com.benasher44.uuid.uuidFrom
+import com.juul.kable.Event.Rejected
+import com.juul.kable.WriteType.WithResponse
+import com.juul.kable.WriteType.WithoutResponse
+import com.juul.kable.external.CLIENT_CHARACTERISTIC_CONFIG_UUID
+import com.juul.kable.gatt.Response.OnCharacteristicRead
+import com.juul.kable.gatt.Response.OnCharacteristicWrite
+import com.juul.kable.gatt.Response.OnDescriptorRead
+import com.juul.kable.gatt.Response.OnDescriptorWrite
+import com.juul.kable.gatt.Response.OnReadRemoteRssi
+import com.juul.kable.gatt.Response.OnServicesDiscovered
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.updateAndGet
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.CoroutineStart.LAZY
+import kotlinx.coroutines.CoroutineStart.UNDISPATCHED
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+
+private val clientCharacteristicConfigUuid = uuidFrom(CLIENT_CHARACTERISTIC_CONFIG_UUID)
+
+internal fun CoroutineScope.peripheral(
+    androidContext: Context,
+    advertisement: Advertisement,
+): Peripheral = peripheral(androidContext, advertisement.bluetoothDevice)
+
+public fun CoroutineScope.peripheral(
+    androidContext: Context,
+    bluetoothDevice: BluetoothDevice,
+): Peripheral = AndroidPeripheral(coroutineContext, androidContext, bluetoothDevice)
+
+public class AndroidPeripheral internal constructor(
+    parentCoroutineContext: CoroutineContext,
+    androidContext: Context,
+    private val bluetoothDevice: BluetoothDevice,
+) : Peripheral {
+
+    private val context = androidContext.applicationContext
+
+    private val job = SupervisorJob(parentCoroutineContext[Job]).apply {
+        invokeOnCompletion { _connection?.close() }
+    }
+    private val scope = CoroutineScope(parentCoroutineContext + job)
+
+    private val _state = MutableStateFlow<State>(State.Disconnected())
+    public override val state: Flow<State> = _state.asStateFlow()
```

Comment:
```suggestion
    public override val state: StateFlow<State> = _state.asStateFlow()
```

Also a Q: what do you get out of the `.asStateFlow()`, does that wrap it or something to prevent checking class?

Reply:
Unless there is a good reason, I'm going to keep this as-is for now. I can open it up (i.e. "change to `StateFlow`" later) and it will be a compatible change, but I can't go the other way and have it be compatible.

`StateFlow` _probably_ is a better choice though, so I suspect the change will be made; but I'm leaning towards a super restrictive API this early on.

Reply:
`.asStateFlow()` returns an immutable `Flow`, which you can't cast to get around that safety.

Reply:
That makes me so happy. Solves a big problem I had with LiveData